### PR TITLE
[BugFix] Avoid unsafe MergeIfStmt across condition writes

### LINKED ISSUE #2538: [BUG][Fuzzer][wrong-code] `MergeIfStmt` silently merges same-condition `if`s that the first body invalidates, giving wrong results
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11

### System information

NVIDIA H100 80GB HBM3 (sm_90), CUDA 13.0, Python 3.11, from-source `v0.1.11` (`cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`). The defect is target-independent — it is in the `MergeIfStmt` TIR transform, applied unconditionally in the CUDA lowering pipeline ([`pipeline.py#L241`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/cuda/pipeline.py#L241)).

### Problem description

`MergeIfStmt` merges two consecutive same-condition `if` statements into one, but doesn't check whether the first body writes a variable the shared condition reads. When it does, the merge is unsound: the second guard should now be false, but the merged form evaluates the condition only once (against the stale value), so the second body runs anyway — a silently wrong result, no error.

```python
Out[0] = 0
if A[0] == 0:
    A[0] = 7      # first body writes the guard variable A[0]
if A[0] == 0:     # A[0] is now 7 -> this guard is false, body must be skipped
    Out[0] = 20
```

With `A[0] == 0` initially, the first `if` fires and sets `A[0] = 7`, so the second guard is false and `Out[0]` must stay `0`. Instead `MergeIfStmt` collapses the two into `if (A[0] == 0) { A[0] = 7; Out[0] = 20; }` — the guard is tested once against the original `A[0]`, both bodies run, and `Out[0]` becomes `20`.

The trigger is specifically *the first body writing the guard variable*, not merging itself: when the first body writes a **different** element (`A[1] = 7`), the guard is not invalidated, the merge is sound, and the result is correct (control below).

### Reproducible example code

```python
import torch
import tilelang
import tilelang.language as T


@tilelang.jit
def kernel(A: T.Tensor((4,), "int32"), Out: T.Tensor((4,), "int32")):
    with T.Kernel(1, threads=1) as bx:
        Out[0] = 0
        if A[0] == 0:
            A[0] = 7          # first body writes the guard variable
        if A[0] == 0:         # now false -> this body must be skipped
            Out[0] = 20


A = torch.zeros(4, dtype=torch.int32, device="cuda")
Out = torch.empty(4, dtype=torch.int32, device="cuda")
kernel(A, Out)
torch.cuda.synchronize()

print("GPU Out[0] =", int(Out[0].cpu()))   # -> 20  (WRONG; correct is 0)
```

Inspecting just the transform (no GPU needed) isolates the collapse to this pass:

```python
from tvm.ir import IRModule

mod = IRModule.from_expr(kernel.prim_func)
print(tilelang.transform.Simplify()(mod)["main"])      # two separate `if A[0]==0` guards (correct)
print(tilelang.transform.MergeIfStmt()(mod)["main"])   # single `if A[0]==0 { A[0]=7; Out[0]=20 }`
```

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

`Out[0]` should be `0`: the second guard is false once the first body sets `A[0] = 7`. `MergeIfStmt` should keep two same-condition `if`s separate (or otherwise preserve sequential semantics) whenever an earlier merged body writes a variable the shared condition reads, since the two guards are no longer equivalent at the point each is evaluated.

### Additional context

**Root cause.** When merging consecutive `IfThenElse` (no `else`) that share a condition, the pass compares the next `if`'s condition to the running `current_condition` with `ExprDeepEqual()` and, on a match, appends the next then-body to the same merged guard ([`merge_if_stmt.cc#L67-L70`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/merge_if_stmt.cc#L67-L70)):

```cpp
if (current_condition.defined() &&
    ExprDeepEqual()(current_condition, if_node->condition)) {
  current_if_bodies.push_back(if_node->then_case);
  continue;
}
```

There is no check that an already-merged body writes a variable the shared condition depends on. The generated CUDA for the repro shows the collapse directly:

```cpp
extern "C" __global__ void main_kernel(int* __restrict__ A, int* __restrict__ Out) {
  Out[0] = 0;
  if (A[0] == 0) {
    A[0] = 7;
    Out[0] = 20;     // second body folded into the first guard
  }
}
```

<details>
<summary>Observed vs expected, and the isolating controls</summary>

```
input A                = [0, 0, 0, 0]
GPU result  Out[0]     = 20      (WRONG)
sequential reference   = 0       (correct: second guard is false after A[0]=7)
```

- `tilelang.transform.MergeIfStmt()` collapses the two guards into one (`IfThenElse` count 2 → 1): `if (A[0] == 0) { A[0] = 7; Out[0] = 20 }`.
- `tilelang.transform.Simplify()` leaves the two guards **separate** (count stays 2) — correct. So the fault is in `MergeIfStmt`, not the surrounding lowering.
- Control: when the first body writes a different element (`A[1] = 7`), the guard `A[0] == 0` is not invalidated, the merge is safe, and the result is correct.

</details>

**Suggested fix.** Before appending a later then-body to a running merged guard at [`merge_if_stmt.cc#L67-L70`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/merge_if_stmt.cc#L67-L70), one option is to require that none of the already-merged bodies writes a variable the shared condition reads (the merged bodies' write-set is disjoint from the condition's read-set); if that does not hold, flushing the current group and starting a new guard (the `else` branch at [L71-L86](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/merge_if_stmt.cc#L71-L86) instead of the `continue`) preserves sequential semantics while keeping the no-op merging that the existing test relies on.

**Provenance.** This merge feature was already bug-fixed once for soundness: [#876](https://github.com/tile-ai/tilelang/pull/876) (merged 2025-09-25) replaced `StructuralEqual` with `ExprDeepEqual` "when merge consecutive If stmt", to avoid merging conditions that only *look* equal. That fix addressed *how conditions are compared*; the *write-invalidates-guard* hole — a body that mutates the condition between the two `if`s — was never addressed. The existing test only exercises the safe shape: [`test_merge_if`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/testing/python/issue/test_tilelang_issue_merge_if.py#L27-L31) checks four guards on *distinct* buffers (`A`, `B`, `C`, `D`), each body writing a constant that does not change its own guard, so the same-condition / first-body-writes-the-guard case is currently untested.

I searched the open and closed issue tracker and did not find an existing report of this defect.

### PR DESCRIPTION
**Pull request by Chennesxu**
## Summary

`MergeIfStmt` merges consecutive `if` statements without `else` when their conditions are structurally equal. This is unsafe when an earlier `if` body writes storage read by a later condition, because the later condition must be re-evaluated after that write.

## Example

```python
if A[0] == 0:
    A[0] = 7
if A[0] == 0:
    Out[0] = 20
```

Before this fix, the pass could rewrite these statements into a single merged if, reusing the first evaluation of `A[0] == 0`. After the first body sets `A[0] = 7`, the second condition no longer holds, so reusing the stale condition changes program behavior.

## Root cause

The merge check only compared condition structure:

```cpp
ExprDeepEqual()(current_condition, if_node->condition)
```

Structural equality is necessary for this rewrite, but it is not sufficient. Even structurally identical expressions may evaluate differently if an earlier pending body modifies a buffer read by the condition.

The previous dependency check also only considered explicit `BufferStore` and `BufferLoad` nodes. Some TileLang operations encode memory accesses through calls such as `tl.access_ptr`, which may later be lowered to `tvm_access_ptr`. After lowering, the original `BufferLoad` is represented by a buffer data Var, offset, extent, and read/write mask, so a scan limited to `BufferLoad` nodes can miss the access.

## Fix

Add a conservative buffer-level dependency guard before merging same-condition if statements:

* collect buffer data vars written by pending if bodies;
* collect buffer data vars read by the shared condition;
* skip the merge when the two sets overlap.

Use a shared visitor for body and condition analysis. It recognizes:

  * explicit `BufferStore` writes;
  * explicit `BufferLoad` reads;
  * call-based memory accesses (e.g., atomic operations, represented as `tl.access_ptr` op before lowering);
  * lowered `tvm_access_ptr` calls;
  * `address_of` passed through opaque calls.

For `tl.access_ptr` and `tvm_access_ptr`, the visitor uses the read/write mask (`1 = read`, `2 = write`, `3 = read/write`). A non-constant mask is conservatively treated as both read and write. Since `address_of` has no access mask, it is also conservatively treated as potentially read and written by its enclosing call.

Buffer data vars are tracked by TVM object identity, so accesses through buffers that share the same underlying data var are treated as accesses to the same storage.

## Scope

The dependency check is intentionally buffer-level rather than index-level. If a pending body writes the same underlying buffer that the condition reads, the pass skips the merge even when the accessed indices may be disjoint.

This keeps the rewrite safe without adding alias or index-range analysis to `MergeIfStmt`. A follow-up may make the optimization less conservative by proving simple disjoint-index cases.

## Tested

Added IR regression coverage in `testing/python/issue/test_tilelang_issue_merge_if.py` for:

* an explicit write to the condition buffer remaining unmerged;
* a safe explicit write to another buffer still merging and preserving both stores;
* an atomic write to the condition buffer remaining unmerged;
* an atomic read in the condition observing an earlier explicit write;
* an atomic write to an unrelated buffer still merging and preserving both bodies;
* an opaque `address_of` call conservatively preventing the merge.

The atomic read/write cases are checked both before and after `LowerAccessPtr`, covering the frontend `tl.access_ptr` and lowered `tvm_access_ptr` representations.

Local verification:

* 7 passed in `testing/python/issue/test_tilelang_issue_merge_if.py`;
* all relevant pre-commit hooks passed.

Verified locally on a TITAN RTX (SM75).

Fixes #2538

### CODE REVIEW

File: src/transform/merge_if_stmt.cc [hunk 1]
```diff
@@ -45,6 +45,40 @@ class MergeIfStmtRewriter : public StmtExprMutator {
...
+        for (const Var &written : written_buffers) {
+          if (load->buffer->data.same_as(written)) {
+            reads_written = true;
+            return;
+          }
+        }
+      }
+    });
+    return reads_written;
+  }
```
[the remark this answers is not in the document]
Comment (Chennesxu) [reply src/transform/merge_if_stmt.cc:143]: Thanks. This PR intentionally keeps the guard local and covers the normal TIR BufferStore/BufferLoad case reported in #2538. I agree opaque calls through access_ptr would require broader access-effect analysis, but wiring that into MergeIfStmt would be a larger follow-up. The current guard is still strictly more conservative for explicit buffer accesses and fixes the reported incorrect merge.

**Comment by Chennesxu** [comment]
> The buffer-level dependency guard is the right level of abstraction here — too conservative would miss valid merges, too fine-grained would require alias analysis that's disproportionate for this pass.
> 
> The approach of collecting written vars from pending `if` bodies and checking overlap with the candidate condition's read vars is clean and easy to reason about. One question: does this guard handle the case where the condition reads a buffer that's written in a _sibling_ `if` body (not a pending merged body)? I think the current code only checks pending merged bodies, which seems correct since sibling `if`s don't share state before merging.
> 
> Tests cover the key cases. Nice fix.

  @muhamedfazalps Thanks for the review!

  You're right — the guard only checks `current_if_bodies` (pending bodies with the same condition), not sibling ifs. This is correct: we only prevent merging when already-accumulated bodies within the same
  condition group would invalidate the condition. Sibling ifs with different conditions are not part of the pending merge group being checked here.
